---
title: "Introduction to "
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(devil)
library(scRNAseq)
```

This vignette provides a step-by-step guide to using the devil package for differential expression analysis of single-cell RNA sequencing (scRNA-seq) data. We’ll start by loading a dataset, preparing the data, fitting the model, and then testing for differential expression between conditions. Finally, we’ll visualize the results using a volcano plot.

## Input data

First, we load a single-cell RNA dataset from the `scRNAseq` package. This dataset contains RNA counts for individual cells and metadata describing the features of each cell.

```{r, message=FALSE, warning=FALSE}
data <- scRNAseq::ReprocessedFluidigmData()
```

### Extracting Counts and Metadata

To fit a model using `devil`, we need to extract the RNA counts and the associated metadata.

```{r}
counts <- data@assays@data[[1]]
metadata <- data@colData

print(dim(counts))  # Dimensions of the counts matrix
print(dim(metadata))  # Dimensions of the metadata
print(head(metadata))  # Display the first few rows of metadata
colnames(metadata)  # Display the column names of metadata
```

The output shows that we have 130 cells described by 28 features, along with RNA counts for 26,255 genes.

### Preprocessing: Filtering Non-Expressed Genes

Before proceeding, let’s remove genes that are not expressed in any of the cells. This helps in reducing the noise and computational load.

```{r}
counts <- counts[rowSums(counts) > 0,]
dim(counts)
```

After filtering, we retain approximately 17,000 genes.

### Creating the Design Matrix

Next, we create a `design_matrix`, which is a matrix that includes the features of each cell. The `model.matrix` function is an easy way to generate this matrix, requiring a formula and the metadata as input.

```{r}
design_matrix <- model.matrix(~Biological_Condition, data = metadata)
print(unique(metadata$Biological_Condition))  # Display unique biological conditions
head(design_matrix)  # Display the first few rows of the design matrix
```

Here, the design matrix models each cell based on its `Biological_Condition`. The intercept in this model represents the “GW16” condition.

## Fitting the model

With the data prepared, we can now fit the model using the `fit_devil` function. This function estimates the coefficients (`beta`) and overdispersion for each gene across different conditions.

```{r}
fit <- devil::fit_devil(
  as.matrix(counts), 
  design_matrix, 
  overdispersion = T, 
  size_factors = T, 
  verbose = T, 
  parallel.cores = 1
)
```

The `fit` object contains several important components, including:

- `beta` : matrix of coefficients (n_genes x n_features)
- `overdispersion` : vector of coefficients (n_genes)

## Testing for Differential Expression

In order to test the data you need to specify your null hypothesis using a contrast vector $c$. Considering a gene $g$ along with its inferred coefficient $\beta_g$, the null hypothesis $H_0$ is usually defined as

$$ H_0 : \sum_i c_i \beta_{g,i} = 0$$
For example, if you are interested in the genes that are differentially expressed between the "GW21" and the "NPC" condition, you need to find the genes for which we strongly reject the null hypothesis 

$$ \beta_{GW21} = \beta_{NPC} \hspace{5mm} \rightarrow \hspace{5mm} \beta_{GW21} - \beta_{NPC} = 0$$
which is equivalent to defining the contrast vector $c = (0,1,0,-1)$.
Once the contrast vector is defined, you can test the null hypothesis using the `test_de` function.

```{r}
contrast <- c(0, 1, 0, -1)
test_res <- devil::test_de(fit, contrast, max_lfc = 20)
if (!('name' %in% colnames(test_res))) {
  test_res$name <- as.character( 1:nrow(test_res) )
}
head(test_res)
```

The results contains, for each gene

- `pval` : p-value associated with the statistical test
- `adj_pval` : p-value corrected considering multiple testing
- `lfc` : log2 fold change of gene expression between the tested conditions

## Visualizing the Results with a Volcano Plot

Finally, we can visualize the results using a volcano plot, which highlights genes based on their log-fold change (`lfc`) and adjusted p-values (`adj_pval`). The `plot_volcano` function from devil makes this easy.

```{r, fig.align='center', fig.width=6, fig.height=6}
devil::plot_volcano(test_res, lfc_cut = 1, pval_cut = .05, labels = TRUE, point_size = 2)
```

This plot provides a clear view of the most significant genes, making it easier to identify those that are likely to be biologically relevant.

## Summary

In this vignette, we’ve demonstrated how to:

	1.	Load and preprocess scRNA-seq data.
	2.	Fit a model using the devil package.
	3.	Perform differential expression testing between conditions.
	4.	Visualize the results using a volcano plot.

The `devil` package offers a powerful and flexible framework for analyzing single-cell RNA sequencing data, helping researchers uncover meaningful biological insights.
